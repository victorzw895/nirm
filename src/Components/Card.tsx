import { Component, Show, createEffect, createSignal } from 'solid-js';
export interface CardProps {
  id: number,
  japName: string,
  engName: string,
  poster: string,
  selectAnime: () => void,
  rank: number | null,
  stars: number | null,
}

const [focusAnimeId, setFocusAnimeId] = createSignal<number>()

const Card: Component<CardProps> = (props) => {
  let cardRef: HTMLDivElement | undefined;

  createEffect(() => {
    if (!cardRef) return;
    if (focusAnimeId() === props.id) {
      cardRef.scrollIntoView({
        behavior: 'smooth',
        block: 'start',
      })
    }
  })

  return (
    <div ref={cardRef} data-testid='card' class='flex items-center bg-dark' onClick={props.selectAnime}>
      <img class='max-h-16 m-2 flex-grow-0' src={props.poster}/>
      <div class='mx-2 overflow-hidden'>
        <p data-testid='jap-name' class='truncate'>{props.japName}</p>
        <p data-testid='eng-name' class='text-xs truncate'>{props.engName}</p>
      </div>
      <Show when={props.rank}>
        <div class='ml-auto mr-3 flex-grow-0'>
          <p data-testid='rank' class='text-3xl'>{props.rank}</p>
          <p data-testid='stars' class='text-xs'>{props.stars ? props.stars : 0}</p>
        </div>
      </Show>
    </div>
  )
}

export { setFocusAnimeId };

export default Card;