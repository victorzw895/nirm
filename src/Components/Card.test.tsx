import { fireEvent, render } from "@solidjs/testing-library";
import Card from "./Card";

const cardProps = {
  id: 0,
  selectAnime: () => {},
  japName: 'jap-name-text',
  engName: 'eng-name-text',
  poster: 'poster-url',
  rank: null,
  stars: null,
};

describe("<Card />", () => {
  it("Displays Jap name", () => {
    const { getByTestId, unmount } = render(() => <Card {...cardProps} />);
    const japName = getByTestId("jap-name");
    expect(japName).toBeInTheDocument();
    expect(japName).toHaveTextContent(/jap-name-text/);
    unmount();
  });

  it("Displays Eng name", () => {
    const { getByTestId, unmount } = render(() => <Card {...cardProps} />);
    const engName = getByTestId("eng-name");
    expect(engName).toBeInTheDocument();
    expect(engName).toHaveTextContent(/eng-name-text/);
    unmount();
  });

  it("Does not displays Rank and stars if rank is null", () => {
    const { queryByTestId, unmount } = render(() => <Card {...cardProps} />);
    const rank = queryByTestId("rank");
    const stars = queryByTestId("stars");
    expect(rank).not.toBeInTheDocument();
    expect(stars).not.toBeInTheDocument();
    unmount();
  });

  it("Displays Rank and stars if rank is not null", () => {
    const { queryByTestId, unmount } = render(() => <Card {...{...cardProps, rank: 1}} />);
    const rank = queryByTestId("rank");
    const stars = queryByTestId("stars");
    expect(rank).toBeInTheDocument();
    expect(rank).toHaveTextContent(/1/);
    expect(stars).toBeInTheDocument();
    expect(stars).toHaveTextContent(/0/);
    unmount();
  });

  it("Opens a preview of anime Card when clicking Card", () => {
    const onClickMock = vi.fn();
    const { getByTestId, unmount } = render(() => <Card {...{...cardProps, selectAnime: onClickMock}} />);
    const card = getByTestId("card");
    fireEvent.click(card);
    expect(onClickMock).toHaveBeenCalledTimes(1);
    unmount();
  })

  it("renders Card", () => {
    const { container, unmount } = render(() => <Card {...cardProps} />);
    expect(container).toMatchSnapshot();
    unmount();
  });
});