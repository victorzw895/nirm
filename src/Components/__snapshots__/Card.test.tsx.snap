// Vitest Snapshot v1

exports[`<Card /> > renders Card 1`] = `
<div>
  <div
    class="flex items-center bg-dark"
    data-testid="card"
  >
    <img
      class="max-h-16 m-2 flex-grow-0"
      src="poster-url"
    />
    <div
      class="mx-2 overflow-hidden"
    >
      <p
        class="truncate"
        data-testid="jap-name"
      >
        jap-name-text
      </p>
      <p
        class="text-xs truncate"
        data-testid="eng-name"
      >
        eng-name-text
      </p>
    </div>
    
  </div>
</div>
`;
